File: server-laravel-single/app/Models/Scopes/ProjectScope.php
<?php

namespace App\Models\Scopes;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Scope;

/**
 * ProjectScope — single-tenant ownership isolation.
 *
 * Auto-discovered by Rhino's HasAutoScope trait (naming convention
 * {Model}Scope). Applied as a global scope on every Project query, so the
 * authenticated user only ever sees / mutates their own projects.
 *
 * Skipped on the console (artisan/seeders) so seeding can create rows for any
 * user, and skipped when unauthenticated (auth:sanctum middleware handles 401).
 */
class ProjectScope implements Scope
{
    public function apply($builder, $model): void
    {
        if (app()->runningInConsole() && !app()->runningUnitTests()) {
            return;
        }

        $user = auth('sanctum')->user() ?? auth()->user();

        if ($user) {
            $builder->where('projects.user_id', $user->getKey());
        }
    }
}

File: server-rails-single/app/models/scopes/project_scope.rb
# frozen_string_literal: true

module Scopes
  # ProjectScope — single-tenant ownership isolation.
  #
  # Auto-discovered by HasAutoScope ({Model}Scope convention) and applied as a
  # default scope on every Project query, so the authenticated user only ever
  # sees / mutates their own projects.
  #
  # With no request user (console/seeds) `user` is nil and no filter is applied,
  # so seeding can create projects for any user.
  class ProjectScope < Rhino::ResourceScope
    def apply(relation)
      return relation unless user

      relation.where(user_id: user.id)
    end
  end
end
